### PR TITLE
feat(cilium-cli-clustermesh): Improve --destination-context option for connecting multiple remote contexts

### PR DESCRIPTION
**Pull request by littlejo**
Improve the `--destination-context` option to enhance the flexibility of connecting multiple remote contexts.

This feature supports various connection modes (using the `--connection-mode` option):

* unicast: configures only the source context
Unicast

Example for connection of 4 clusters:
```
cilium clustermesh connect --connection-mode=unicast --context=kind-cmesh1 --destination-context=kind-cmesh2,kind-cmesh3,kind-cmesh4
✨ Extracting access information of cluster cmesh1...
🔑 Extracting secrets from cluster cmesh1...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.5]
✨ Extracting access information of cluster cmesh2...
🔑 Extracting secrets from cluster cmesh2...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.3]
⚠️ Cilium CA certificates do not match between clusters. Multicluster features will be limited!
✨ Extracting access information of cluster cmesh3...
🔑 Extracting secrets from cluster cmesh3...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.2]
⚠️ Cilium CA certificates do not match between clusters. Multicluster features will be limited!
✨ Extracting access information of cluster cmesh4...
🔑 Extracting secrets from cluster cmesh4...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.4]
⚠️ Cilium CA certificates do not match between clusters. Multicluster features will be limited!
ℹ️ Configuring Cilium in cluster kind-cmesh1 to connect to cluster kind-cmesh2,kind-cmesh3,kind-cmesh4
✅ Connected cluster kind-cmesh1 => kind-cmesh2!
✅ Connected cluster kind-cmesh1 => kind-cmesh3!
✅ Connected cluster kind-cmesh1 => kind-cmesh4!
```
Example for disconnection of 4 clusters:

```
cilium clustermesh disconnect --connection-mode=unicast --context=kind-cmesh1 --destination-context=kind-cmesh2,kind-cmesh3,kind-cmesh4
ℹ️ Configuring Cilium in cluster kind-cmesh1 to disconnect from cluster kind-cmesh2,kind-cmesh3,kind-cmesh4
✅ Disconnected clusters kind-cmesh1 x> kind-cmesh2!
✅ Disconnected clusters kind-cmesh1 x> kind-cmesh3!
✅ Disconnected clusters kind-cmesh1 x> kind-cmesh4!
```

* bidirectional (default value): configures the source context and its cluster in the destination contexts
bidirectional

Example for connection of 4 clusters:

```
cilium clustermesh connect --connection-mode=bidirectional --context=kind-cmesh1 --destination-context=kind-cmesh2,kind-cmesh3,kind-cmesh4
✨ Extracting access information of cluster cmesh1...
🔑 Extracting secrets from cluster cmesh1...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.2]
✨ Extracting access information of cluster cmesh2...
🔑 Extracting secrets from cluster cmesh2...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.3]
⚠️ Cilium CA certificates do not match between clusters. Multicluster features will be limited!
✨ Extracting access information of cluster cmesh3...
🔑 Extracting secrets from cluster cmesh3...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.4]
⚠️ Cilium CA certificates do not match between clusters. Multicluster features will be limited!
✨ Extracting access information of cluster cmesh4...
🔑 Extracting secrets from cluster cmesh4...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.5]
⚠️ Cilium CA certificates do not match between clusters. Multicluster features will be limited!
ℹ️ Configuring Cilium in cluster kind-cmesh1 to connect to cluster kind-cmesh2,kind-cmesh3,kind-cmesh4
ℹ️ Configuring Cilium in cluster kind-cmesh2 to connect to cluster kind-cmesh1
ℹ️ Configuring Cilium in cluster kind-cmesh3 to connect to cluster kind-cmesh1
ℹ️ Configuring Cilium in cluster kind-cmesh4 to connect to cluster kind-cmesh1
✅ Connected cluster kind-cmesh1 <=> kind-cmesh2!
✅ Connected cluster kind-cmesh1 <=> kind-cmesh3!
✅ Connected cluster kind-cmesh1 <=> kind-cmesh4!
```

Example for disconnection of 4 clusters:

```
cilium clustermesh disconnect --connection-mode=bidirectional --context=kind-cmesh1 --destination-context=kind-cmesh2,kind-cmesh3,kind-cmesh4
ℹ️ Configuring Cilium in cluster kind-cmesh1 to disconnect from cluster kind-cmesh2,kind-cmesh3,kind-cmesh4
ℹ️ Configuring Cilium in cluster kind-cmesh2 to disconnect from cluster kind-cmesh1
ℹ️ Configuring Cilium in cluster kind-cmesh3 to disconnect from cluster kind-cmesh1
ℹ️ Configuring Cilium in cluster kind-cmesh4 to disconnect from cluster kind-cmesh1
✅ Disconnected clusters kind-cmesh1 <x> kind-cmesh2!
✅ Disconnected clusters kind-cmesh1 <x> kind-cmesh3!
✅ Disconnected clusters kind-cmesh1 <x> kind-cmesh4!
```

* mesh: configures both the source and destination contexts, including clusters in both.
Mesh

Example for connection of 4 clusters:

```
cilium clustermesh connect --connection-mode=mesh --context=kind-cmesh1 --destination-context=kind-cmesh2,kind-cmesh3,kind-cmesh4
✨ Extracting access information of cluster cmesh1...
🔑 Extracting secrets from cluster cmesh1...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.2]
✨ Extracting access information of cluster cmesh2...
🔑 Extracting secrets from cluster cmesh2...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.5]
⚠️ Cilium CA certificates do not match between clusters. Multicluster features will be limited!
✨ Extracting access information of cluster cmesh3...
🔑 Extracting secrets from cluster cmesh3...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.4]
⚠️ Cilium CA certificates do not match between clusters. Multicluster features will be limited!
✨ Extracting access information of cluster cmesh4...
🔑 Extracting secrets from cluster cmesh4...
⚠️  Service type NodePort detected! Service may fail when nodes are removed from the cluster!
ℹ️  Found ClusterMesh service IPs: [172.18.0.3]
⚠️ Cilium CA certificates do not match between clusters. Multicluster features will be limited!
ℹ️ Configuring Cilium in cluster kind-cmesh1 to connect to cluster kind-cmesh2,kind-cmesh3,kind-cmesh4
ℹ️ Configuring Cilium in cluster kind-cmesh2 to connect to cluster kind-cmesh3,kind-cmesh4,kind-cmesh1
ℹ️ Configuring Cilium in cluster kind-cmesh3 to connect to cluster kind-cmesh2,kind-cmesh4,kind-cmesh1
ℹ️ Configuring Cilium in cluster kind-cmesh4 to connect to cluster kind-cmesh2,kind-cmesh3,kind-cmesh1
✅ Connected clusters kind-cmesh2, kind-cmesh3, kind-cmesh4, and kind-cmesh1!
```
Example for disconnection of 4 clusters:

```
cilium clustermesh disconnect --connection-mode=mesh --context=kind-cmesh1 --destination-context=kind-cmesh2,kind-cmesh3,kind-cmesh4
ℹ️ Configuring Cilium in cluster kind-cmesh1 to disconnect from cluster kind-cmesh2,kind-cmesh3,kind-cmesh4
ℹ️ Configuring Cilium in cluster kind-cmesh2 to disconnect from cluster kind-cmesh3,kind-cmesh4,kind-cmesh1
ℹ️ Configuring Cilium in cluster kind-cmesh3 to disconnect from cluster kind-cmesh2,kind-cmesh4,kind-cmesh1
ℹ️ Configuring Cilium in cluster kind-cmesh4 to disconnect from cluster kind-cmesh2,kind-cmesh3,kind-cmesh1
✅ Disconnected clusters kind-cmesh2, kind-cmesh3, kind-cmesh4, and kind-cmesh1!
```

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

### CODE REVIEW

**Review by thorn3r — CHANGES_REQUESTED** [review]
Thanks for the contribution! This is a really nice convenience enhancement.

It looks like there are some tests breaking on the changes you made to existing funcs. Could you update those and also add some testing around the new `connection-mode` and `destination-contexts` you added?

**Comment by littlejo** [comment]
@thorn3r do you have some news?

File: cilium-cli/cli/clustermesh.go
Comment (derailed) [inline cilium-cli/cli/clustermesh.go]: This seems a bit confusing ie is destination-context still valid or is it destination-contexts=xxx? 
Not sure we need to keep both options??

File: cilium-cli/clustermesh/clustermesh.go
Comment (derailed) [inline cilium-cli/clustermesh/clustermesh.go]: Would be best to check for err first prior to assignment

File: cilium-cli/clustermesh/clustermesh.go
Comment (derailed) [inline cilium-cli/clustermesh/clustermesh.go]: Lots going on here... Not too keen about all this state and assignments. 
Is there a way we can clean this up a bit using functions to gather the necessary state?

Also depending on the connection type some of the modes may not require all that pre-computed state?
Lastly, breaking this up in sep func/methods would make testing this functionality easier.

File: cilium-cli/clustermesh/clustermesh.go
Comment (derailed) [inline cilium-cli/clustermesh/clustermesh.go]: Do we need to continue here or should this be an error?

File: cilium-cli/clustermesh/clustermesh.go
Comment (derailed) [inline cilium-cli/clustermesh/clustermesh.go]: Aren't those mutually exclusive? if so it might be best to switch based on mode. 
Also if bd, unicast, etc.. are interesting it would be great to use constants

File: cilium-cli/clustermesh/clustermesh.go
Comment (derailed) [inline cilium-cli/clustermesh/clustermesh.go]: I think slices.DeleteFunc could be used vs our own?

File: cilium-cli/clustermesh/clustermesh.go
Comment (derailed) [inline cilium-cli/clustermesh/clustermesh.go]: Empty slice? Prefer var remoteHmValsAll []map[string]interface{}

File: cilium-cli/clustermesh/clustermesh.go
Comment (derailed) [inline cilium-cli/clustermesh/clustermesh.go]: Do we need to check bounds here?

File: cilium-cli/clustermesh/clustermesh.go
Comment (derailed) [inline cilium-cli/clustermesh/clustermesh.go]: Here and above... Do we need to check slice bounds prior to indexing?

File: cilium-cli/clustermesh/clustermesh_test.go
Comment (derailed) [inline cilium-cli/clustermesh/clustermesh_test.go:21]: Perhaps slices.EqualFunc might help here

**Review by derailed — CHANGES_REQUESTED** [review]
@littlejo Thank you for this update! 
I think this is great but needs a little refactor + TLC

File: cilium-cli/cli/clustermesh.go
Comment (littlejo) [reply cilium-cli/cli/clustermesh.go]: For retro-compatibility, I kept destination-context. You can use both, it just add to the destination-contexts list. But if you think that the destination-context option is now useless. I can remove this option.

File: cilium-cli/clustermesh/clustermesh.go
Comment (littlejo) [reply cilium-cli/clustermesh/clustermesh.go]: > Do we need to continue here or should this be an error?

We need to continue. It's just a warning. It's already on the main branch:
https://github.com/cilium/cilium/blob/main/cilium-cli/clustermesh/clustermesh.go#L1623

File: cilium-cli/clustermesh/clustermesh.go
Comment (littlejo) [reply cilium-cli/clustermesh/clustermesh.go]: Thank you. i didn't know.

File: cilium-cli/cli/clustermesh.go
Comment (littlejo) [reply cilium-cli/cli/clustermesh.go]: I removed destination-contexts option and i modify destination-context option to have more contexts

**Review by thorn3r — APPROVED** [review]
Improvement of `destination-context` rather than new flag feels like the right call :+1: Nice work!

**Comment by littlejo** [comment]
@dylandreimerink Could you /test again ?

**Comment by littlejo** [comment]
@dylandreimerink Could you /test again ?

**Comment by littlejo** [comment]
Hello @dylandreimerink and @thorn3r , i saw 3 errors but i think no one about this improvement.

**Comment by dylandreimerink** [comment]
> i saw 3 errors but i think no one about this improvement.

Agreed, I have triggered retires, some of these tests are a bit flaky unfortunately.  Its mostly about the `ci-e2e-upgrade` test, it is required, the other two are not.

Also blocking this PR from merging is that we require a new reviews from @derailed

**Review by derailed — APPROVED** [review]
@littlejo Thank you for the updates! I think we should perhaps make a note here to deprecate destinationContext->destinationContexts as it is now a bit of a misnomer.

**Pull request merged**
